### PR TITLE
typescript: add logs to traces

### PR DESCRIPTION
**Pull request by fredr**
Adds logs rows to traces in typescript

### CODE REVIEW

File: runtimes/core/src/model/mod.rs [hunk 1]
```diff
@@ -288,3 +288,55 @@ pub struct AuthSuccessResponse {
...
+impl From<crate::log::LogLevel> for LogLevel {
+    fn from(level: crate::log::LogLevel) -> Self {
+        match level {
+            crate::log::LogLevel::Trace => LogLevel::Trace,
+            crate::log::LogLevel::Debug => LogLevel::Debug,
+            crate::log::LogLevel::Info => LogLevel::Info,
+            crate::log::LogLevel::Warn => LogLevel::Warn,
+            crate::log::LogLevel::Error => LogLevel::Error,
+
+            // TODO(fredr): these are not supported by the go runtime
```
Comment (fredr) [inline runtimes/core/src/model/mod.rs:337]: Not sure how we should handle this :thinking:

File: runtimes/core/src/log/logger.rs [hunk 2]
```diff
@@ -202,6 +212,8 @@ impl Logger {
             }
         }
 
+        self.write_to_trace(request, level, &msg, &values);
```
Comment (eandre) [inline runtimes/core/src/log/logger.rs:215]: We're inserting the message as a value above. Do we need to pass it in here, too? It requires cloning the message, per above.

File: runtimes/core/src/log/logger.rs [hunk 3]
```diff
@@ -235,6 +247,64 @@ impl Logger {
...
+                        key,
+                        value: model::LogFieldValue::I64(num.as_i64().unwrap()),
+                    });
+                } else if num.is_u64() {
+                    trace_fields.push(LogField {
+                        key,
+                        value: model::LogFieldValue::U64(num.as_u64().unwrap()),
+                    });
+                } else if num.is_f64() {
+                    println!("is_f64");
```
Comment (eandre) [inline runtimes/core/src/log/logger.rs:274]: Remove

File: runtimes/core/src/model/mod.rs [hunk 4]
```diff
@@ -288,3 +288,55 @@ pub struct AuthSuccessResponse {
     /// The user data.
     pub user_data: serde_json::Map<String, serde_json::Value>,
 }
+
+// matches go runtime
+pub enum LogLevel {
```
Comment (eandre) [inline runtimes/core/src/model/mod.rs:293]: In the logging overhaul PR I removed the log level enum we had and relied on `::log::Level` instead. Maybe better to do so here, too?

File: runtimes/core/src/trace/protocol.rs [hunk 5]
```diff
@@ -57,6 +57,57 @@ impl Tracer {
...
+pub struct LogMessageData<'a> {
+    pub source: Option<&'a Request>,
+    pub msg: &'a str,
+    pub level: LogLevel,
+    pub fields: Vec<LogField<'a>>,
+}
+
+impl Tracer {
+    #[inline]
+    pub fn log_message(&self, data: LogMessageData) -> Option<TraceEventId> {
```
Comment (eandre) [inline runtimes/core/src/trace/protocol.rs:69]: I think we don't need to return an event id. It's mainly used for log events that need to correlate "start" and "end", like beginning a database query and recording its completion.

**Review by eandre — COMMENTED** [review]
Overall this looks great! A few minor nitpicks :)

File: runtimes/core/src/log/logger.rs [hunk 6]
```diff
@@ -236,6 +248,65 @@ impl Logger {
...
+
+        let _ = self
+            .tracer
+            .read()
+            .expect("tracer lock poisoned")
+            .log_message(LogMessageData {
+                source: request,
+                msg,
+                level,
+                fields: trace_fields,
```
Comment (eandre) [inline runtimes/core/src/log/logger.rs:307]: Instead of allocating a vec, could we pass in an iterator instead?

**Review by eandre — APPROVED** [review]

**Pull request merged**